r/R/filesystem.R

```diff
@@ -242,11 +242,31 @@ FileSystem <- R6Class("FileSystem", inherit = ArrowObject,
   )
 )
 FileSystem$from_uri <- function(uri) {
+  assert_that(is.string(uri))
   out <- fs___FileSystemFromUri(uri)
   out$fs <- shared_ptr(FileSystem, out$fs)$..dispatch()
   out
 }
 
+get_path_and_filesystem <- function(x, filesystem = NULL) {
+  # Wrapper around FileSystem$from_uri that handles local paths
+  # and an optional explicit filesystem
+  assert_that(is.string(x))
+  if (is_url(x)) {
+    if (!is.null(filesystem)) {
+      # Stop? Can't have URL (which yields a fs) and another fs
```

Comment:
Indeed, the user shouldn't pass both a filesystem object and a URI, as then things like `filesystem$OpenInputFile(uri)` will fail (that was the root cause of the python parquet s3 test crashes -> https://github.com/apache/arrow/pull/8139). So I think can either let it pass here and rely on a later error in eg OpenInputFile, or otherwise raise a (potentially more informative) error message here.

Reply:
What happens here is that `filesystem` will be ignored if `x` is a URL, i.e. we'll use the filesystem that `from_uri(x)` returns. So that should either Just Work if the filesystems happen to be the same, or you're right, it should error with some kind of File Not Found later.